### ISSUE TITLE
master node fails to start with >= 2.3.1

### ISSUE THREAD

**Issue by keien** [aws/aws-parallelcluster#1249]
**Environment:**
 - AWS ParallelCluster / CfnCluster version 2.3.1
 - OS: Centos 7
 - Scheduler: SGE
 - Master instance type: t2.medium

**Bug description and how to reproduce:**

this is likely an issue specific to the way we've provisioned our AMI, but it's preventing us from upgrading beyond 2.1.1, so we'd love to at least get some guidance on how to further debug this issue

during master provisioning with our AMI (built on top of the parallelcluster AMI in the AMI list for the 2.3.1 version), we run into the following issue with chef:
```
    ================================================================================
    Error executing action `install` on resource 'python_package[setuptools]'
    ================================================================================

    Mixlib::ShellOut::ShellCommandFailed
    ------------------------------------
    Expected process to exit with [0], but received '1'
    ---- Begin output of ["/usr/bin/python", "-", "setuptools"] ----
    STDOUT:
    STDERR: From cffi callback <function _verify_callback at 0x7f5b30f3d500>:
    Traceback (most recent call last):
      File "/usr/lib/python2.7/site-packages/OpenSSL/SSL.py", line 309, in wrapper
        _lib.X509_up_ref(x509)
    AttributeError: 'module' object has no attribute 'X509_up_ref'
    From cffi callback <function _verify_callback at 0x7f5b30edd050>:
    Traceback (most recent call last):
      File "/usr/lib/python2.7/site-packages/OpenSSL/SSL.py", line 309, in wrapper
        _lib.X509_up_ref(x509)
    AttributeError: 'module' object has no attribute 'X509_up_ref'
    From cffi callback <function _verify_callback at 0x7f5b30edd1b8>:
    Traceback (most recent call last):
      File "/usr/lib/python2.7/site-packages/OpenSSL/SSL.py", line 309, in wrapper
        _lib.X509_up_ref(x509)
    AttributeError: 'module' object has no attribute 'X509_up_ref'
    From cffi callback <function _verify_callback at 0x7f5b30f0c500>:
    Traceback (most recent call last):
      File "/usr/lib/python2.7/site-packages/OpenSSL/SSL.py", line 309, in wrapper
        _lib.X509_up_ref(x509)
    AttributeError: 'module' object has no attribute 'X509_up_ref'
    From cffi callback <function _verify_callback at 0x7f5b30f0ccf8>:
    Traceback (most recent call last):
      File "/usr/lib/python2.7/site-packages/OpenSSL/SSL.py", line 309, in wrapper
        _lib.X509_up_ref(x509)
    AttributeError: 'module' object has no attribute 'X509_up_ref'
    From cffi callback <function _verify_callback at 0x7f5b30edd2a8>:
    Traceback (most recent call last):
      File "/usr/lib/python2.7/site-packages/OpenSSL/SSL.py", line 309, in wrapper
        _lib.X509_up_ref(x509)
    AttributeError: 'module' object has no attribute 'X509_up_ref'
    No handlers could be found for logger "pip._internal.index"
    Traceback (most recent call last):
      File "<stdin>", line 43, in <module>
      File "/usr/lib/python2.7/site-packages/pip/_internal/index.py", line 543, in find_requirement
        'No matching distribution found for %s' % req
    pip._internal.exceptions.DistributionNotFound: No matching distribution found for setuptools
    ---- End output of ["/usr/bin/python", "-", "setuptools"] ----
    Ran ["/usr/bin/python", "-", "setuptools"] returned 1
```
after the above, I attempted to fix it a couple different ways, and ultimately uninstalling the yum `python2-paramiko` package fixes the issue for the master provisioning

next error I ran into was that the master could not register new compute nodes, due to
```
[ip-172-31-116-160]root:log [02:37]
$ (bash -x /opt/parallelcluster/scripts/compute_ready)

+ . /etc/parallelcluster/cfnconfig
++ stack_name=parallelcluster-debug-ami-0ce9e27878fc7fab6
++ cfn_preinstall=s3://cb-parallelcluster/pre_install.sh
++ cfn_preinstall_args=NONE
++ cfn_postinstall=s3://cb-parallelcluster/post_install.sh
++ cfn_postinstall_args='quota no-hthread other'
++ cfn_region=us-west-2
++ cfn_scheduler=sge
++ cfn_scheduler_slots=cores
++ cfn_instance_slots=36
++ cfn_encrypted_ephemeral=false
++ cfn_ephemeral_dir=/scratch
++ cfn_shared_dir=/shared
++ cfn_proxy=NONE
++ cfn_node_type=ComputeFleet
++ cfn_cluster_user=centos
++ cfn_sqs_queue=https://sqs.us-west-2.amazonaws.com/171496337684/parallelcluster-debug-ami-0ce9e27878fc7fab6-SQS-PIIFP71H7YE3
++ cfn_master=ip-172-31-117-158.us-west-2.compute.internal
+ instance_id_url=http://169.254.169.254/latest/meta-data/instance-id
++ curl --retry 3 --retry-delay 0 --silent --fail http://169.254.169.254/latest/meta-data/instance-id
+ instance_id=i-080a9826604eb14e3
+ local_hostname_url=http://169.254.169.254/latest/meta-data/local-hostname
++ curl --retry 3 --retry-delay 0 --silent --fail http://169.254.169.254/latest/meta-data/local-hostname
+ local_hostname='ip-172-31-116-160.us-west-2.compute.internal cerebras.aws'
+ aws --region us-west-2 sqs send-message --queue-url https://sqs.us-west-2.amazonaws.com/171496337684/parallelcluster-debug-ami-0ce9e27878fc7fab6-SQS-PIIFP71H7YE3 --message-body '{"Type" : "Notification", "Message" : "{\"StatusCode\":\"Complete\",\"Description\":\"Succesfully launched i-080a9826604eb14e3\",\"Event\":\"parallelcluster:COMPUTE_READY\",\"EC2InstanceId\":\"i-080a9826604eb14e3\",\"Slots\":\"36\",\"LocalHostname\":\"ip-172-31-116-160.us-west-2.compute.internal' 'cerebras.aws\"}"}'
usage: aws [options] <command> <subcommand> [<subcommand> ...] [parameters]
To see help text, you can run:

  aws help
  aws <command> help
  aws <command> <subcommand> help

Unknown options: cerebras.aws\"}"}
bash: virtualenv_info: command not found
bash: get_git_branch: command not found
```

this is because in our VPC, `http://169.254.169.254/latest/meta-data/local-hostname` returns two domains, so I have to patch the `/opt/parallelcluster/scripts/compute_ready` during pre-install to get around this (the old 2.1.1 `/opt/parallelcluster/scripts/compute_ready` script seems like it didn't try to use the `local-hostname` endpoint)

next error after that is in sqswatcher:
```
2019-05-07 13:10:03,814 INFO [sqswatcher:main] sqswatcher startup
2019-05-07 13:10:03,814 INFO [sqswatcher:_get_config] Reading /etc/sqswatcher.cfg
2019-05-07 13:10:03,815 INFO [sqswatcher:_get_config] Configured parameters: region=us-west-2 scheduler=sge sqsqueue=parallelcluster-debug-ami-0ce9e27878fc7fab6-SQS-PIIFP71H7YE3 table_name=parallelcluster-debug-ami-0ce9e27878fc7fab6-DynamoDBTable-1VAO9VRMOCYJA cluster_user=centos proxy=NONE max_queue_size=16 stack_name=parallelcluster-debug-ami-0ce9e27878fc7fab6
2019-05-07 13:10:04,093 INFO [utils:get_asg_name] ASG parallelcluster-debug-ami-0ce9e27878fc7fab6-ComputeFleet-ZPAYW2S2OMJ0 found for the stack parallelcluster-debug-ami-0ce9e27878fc7fab6
2019-05-07 13:10:04,093 CRITICAL [sqswatcher:main] An unexpected error occurred: No module named paramiko
2019-05-07 13:10:34,118 INFO [sqswatcher:main] sqswatcher startup
2019-05-07 13:10:34,118 INFO [sqswatcher:_get_config] Reading /etc/sqswatcher.cfg
2019-05-07 13:10:34,118 INFO [sqswatcher:_get_config] Configured parameters: region=us-west-2 scheduler=sge sqsqueue=parallelcluster-debug-ami-0ce9e27878fc7fab6-SQS-PIIFP71H7YE3 table_name=parallelcluster-debug-ami-0ce9e27878fc7fab6-DynamoDBTable-1VAO9VRMOCYJA cluster_user=centos proxy=NONE max_queue_size=16 stack_name=parallelcluster-debug-ami-0ce9e27878fc7fab6
2019-05-07 13:10:34,361 INFO [utils:get_asg_name] ASG parallelcluster-debug-ami-0ce9e27878fc7fab6-ComputeFleet-ZPAYW2S2OMJ0 found for the stack parallelcluster-debug-ami-0ce9e27878fc7fab6
2019-05-07 13:10:34,362 CRITICAL [sqswatcher:main] An unexpected error occurred: No module named paramiko
```

this was because during the uninstall of the `python2-paramiko` package, for some reason it left behind a few files in ` /usr/lib64/python2.7/site-packages` that causes pip to think that the package is already installed during the pcluster chef init

we tried to fix this by cleaning up those files in our base image, but we still ran into the same issue

this is as far as managed to get

**Comment by keien** [comment]
Additional context:

config:
```
[global]
update_check = true
sanity_check = true
cluster_template = m5-4xlarge-ami-08dee0cef576f9b8d-spot

[cluster m5-4xlarge-ami-08dee0cef576f9b8d-spot]
custom_ami = ami-08dee0cef576f9b8d
# below is inserted from config/m5-4xlarge-spot
compute_instance_type = m5.4xlarge
cluster_type = spot
spot_price = 1.00
master_instance_type = t2.2xlarge
master_root_volume_size = 100
compute_root_volume_size = 100
initial_queue_size = 0
max_queue_size = 64
maintain_initial_size = false
base_os=centos7
key_name = ec2-admin
pre_install = s3://cb-parallelcluster/pre_install.sh
post_install = s3://cb-parallelcluster/post_install.sh
post_install_args = quota
extra_json = { "cluster" : { "cfn_scheduler_slots" : "cores" } }
vpc_settings = default
scaling_settings = default
# below is inserted from config/vpc-default
[vpc default]
vpc_id = vpc-fa52d99e
master_subnet_id = subnet-05fd9823327362ff2
vpc_security_group_id = sg-74e42c12
use_public_ips = false
# below is inserted from config/scaling-profiles
[scaling default]
# Threshold for triggering CloudWatch ScaleUp action
# (defaults to 1 for default template)
scaling_threshold = 1
# Number of instances to add when called CloudWatch ScaleUp action
# (defaults to 1 for default template)
scaling_adjustment = 1
# Threshold for triggering CloudWatch ScaleUp2 action
# (defaults to 200 for default template)
#scaling_threshold2 = 200
# Number of instances to add when called CloudWatch ScaleUp2 action
# (defaults to 20 for default template)
#scaling_adjustment2 = 20
# Period to measure ScalingThreshold
# (defaults to 60 for default template)
scaling_period = 30
# Period to measure ScalingThreshold
# (defaults to 2 for default template)
scaling_evaluation_periods = 2
# Amount of time in seconds to wait before attempting further scaling actions
# (defaults to 300 for the default template
scaling_cooldown = 300

# scaling config for aggressive one-job-one-instance burst usage
[scaling one-to-one]
scaling_threshold = 1
scaling_adjustment = 1
scaling_threshold2 = 20
scaling_adjustment2 = 10
```

pre-install:
```
hostnamectl set-hostname `curl -s http://169.254.169.254/latest/meta-data/hostname | cut -f 1 -d' '`

# As of pcluster 2.3.1, we need to do a similar fixup in their compute_ready script which uses the
# above URL to get a hostname
sed -i 's/local_hostname=$(curl --retry 3 --retry-delay 0 --silent --fail ${local_hostname_url})/local_hostname=$(curl --retry 3 --retry-delay 0 --silent --fail ${local_hostname_url} | cut -f 1 -d" ")/' /opt/parallelcluster/scripts/compute_ready
```

**Comment by lukeseawalker** [comment]
Hi,
as part of [ParallelCluster 2.4.1](https://github.com/aws/aws-parallelcluster/releases/tag/v2.4.1) we released a series of enhancements that allow the cluster daemons and cookbook python dependencies to be installed in isolated virtualenvs. This should solve the issues you are having with the packages and their dependencies. Could you try to create a new custom AMI starting from those released with the 2.4.1 and see if the issue still occurs?
About the compute_ready workaround, we currently do not support custom DHCP option set out of the box. I take note to improve the handling of the local-hostname metadata in the case it contains more than one value.

**Comment by keien** [comment]
looks like the 2.4.1 image worked - thanks!
